ql/src/test/org/apache/hadoop/hive/ql/udf/ptf/TestValueBoundaryScanner.java

```diff
@@ -0,0 +1,187 @@
+import org.apache.hadoop.hive.common.type.TimestampTZ;
+import org.apache.hadoop.hive.ql.plan.ptf.OrderExpressionDef;
+import org.apache.hadoop.hive.ql.plan.ptf.PTFExpressionDef;
+import org.apache.hadoop.hive.serde2.io.DateWritableV2;
+import org.apache.hadoop.hive.serde2.io.HiveDecimalWritable;
+import org.apache.hadoop.hive.serde2.io.TimestampLocalTZWritable;
+import org.apache.hadoop.hive.serde2.io.TimestampWritableV2;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestValueBoundaryScanner {
+
+  @Test
+  public void testLongEquals() {
+    PTFExpressionDef argDef = new PTFExpressionDef();
+    argDef.setOI(PrimitiveObjectInspectorFactory.writableLongObjectInspector);
+
+    LongValueBoundaryScanner scanner =
+        new LongValueBoundaryScanner(null, null, new OrderExpressionDef(argDef), false);
+    LongWritable w1 = new LongWritable(1);
+    LongWritable w2 = new LongWritable(2);
+
+    Assert.assertTrue(scanner.isEqual(w1, w1));
+
+    Assert.assertFalse(scanner.isEqual(w1, w2));
+    Assert.assertFalse(scanner.isEqual(w2, w1));
+
+    Assert.assertFalse(scanner.isEqual(null, w2));
+    Assert.assertFalse(scanner.isEqual(w1, null));
+
+    Assert.assertTrue(scanner.isEqual(null, null));
+  }
+
+  @Test
+  public void testHiveDecimalEquals() {
+    PTFExpressionDef argDef = new PTFExpressionDef();
+    argDef.setOI(PrimitiveObjectInspectorFactory.writableHiveDecimalObjectInspector);
+
+    HiveDecimalValueBoundaryScanner scanner =
+        new HiveDecimalValueBoundaryScanner(null, null, new OrderExpressionDef(argDef), false);
+    HiveDecimalWritable w1 = new HiveDecimalWritable(1);
+    HiveDecimalWritable w2 = new HiveDecimalWritable(2);
+
+    Assert.assertTrue(scanner.isEqual(w1, w1));
+
+    Assert.assertFalse(scanner.isEqual(w1, w2));
+    Assert.assertFalse(scanner.isEqual(w2, w1));
+
+    Assert.assertFalse(scanner.isEqual(null, w2));
+    Assert.assertFalse(scanner.isEqual(w1, null));
+
+    Assert.assertTrue(scanner.isEqual(null, null));
+  }
+
+  @Test
+  public void testDateEquals() {
+    PTFExpressionDef argDef = new PTFExpressionDef();
+    argDef.setOI(PrimitiveObjectInspectorFactory.writableDateObjectInspector);
+
+    DateValueBoundaryScanner scanner =
+        new DateValueBoundaryScanner(null, null, new OrderExpressionDef(argDef), false);
+    Date date = new Date();
+    date.setTimeInMillis(1000);
+    DateWritableV2 w1 = new DateWritableV2(date);
+    DateWritableV2 w2 = new DateWritableV2(date);
+    DateWritableV2 w3 = new DateWritableV2(); // empty
+
+    Assert.assertTrue(scanner.isEqual(w1, w2));
+    Assert.assertTrue(scanner.isEqual(w2, w1));
+
+    // empty == epoch
+    Assert.assertTrue(scanner.isEqual(w3, new DateWritableV2(new Date())));
+    // empty == another
+    Assert.assertTrue(scanner.isEqual(w1, w3));
```

Comment:
Is this expected?

Reply:
not sure at the moment, but this worked this way even without my patch

Reply:
ok, I found that, in case of dates, 1000s increment results in the same day :) so testing issue, fixed in
https://github.com/apache/hive/pull/1950/commits/589126936825ea6652787457dcff6ca038d37f6a